### PR TITLE
use apex AMP O1 optimizations

### PR DESCRIPTION
**Pull request by repodiac**

### CODE REVIEW

**Comment by repodiac** [comment]
Hey... not much going on here!? Do you need any help for pursuing the PR merge? Just let me know then, thx.

**Comment by erogol** [comment]
I am not still sure if it's a good idea to have apex as a dependency to the main training script. It causes a bug dependency overhead. Maybe we can create a new train_apex.py wrapping the functions in train.py so that we can decouple it from the main flow. What do you say?

**Comment by repodiac** [comment]
What do you mean with "bug dependency overhead" exactly? I have invested significant efforts in providing the "drop-in solution" for APEX O1 - it is configurable via the config.json file, so why do you have concerns about integrating it? If someone does not want to use it, no problem - just set the config accordingly!
Separating it via an additional train_<something>.py file does not really help I suppose... honestly, I am a bit disappointed if you object to integrate it into the project :-/

**Comment by erogol** [comment]
> 
> 
> What do you mean with "bug dependency overhead" exactly? I have invested significant efforts in providing the "drop-in solution" for APEX O1 - it is configurable via the config.json file, so why do you have concerns about integrating it? If someone does not want to use it, no problem - just set the config accordingly!
> Separating it via an additional train_.py file does not really help I suppose... honestly, I am a bit disappointed if you object to integrate it into the project :-/

Don't get me wrong. This is a great contribution and I appreciate it. I don't also object it whatsoever.

But as you might appreciate this project should work for everyone and support every possible system as much as possible. APEX is a big dependency and it has many requirements in terms of PyTorch, CUDA, and C++ compilers. It limits the range of systems that we can run TTS. 

I just try to find a better solution. If people like to use APEX, we can serve as an option but if they don't we need to get rid of its overhead. So I suggest having a train_apex.py and an option in setup.py to enable APEX if this is what the user desires. 

In your current PR, even if you don't use APEX, still you need to install it since we import it anyway. Maybe one easier option is to import APEX if it is available in the system. 

I hope you see my point.

File: train.py [hunk 1]
```diff
@@ -484,6 +493,14 @@ def main(args):  # pylint: disable=redefined-outer-name
     else:
         optimizer_st = None
 
+    if c.apex_amp_level:
+        # pylint: disable=import-outside-toplevel
+        from apex import amp
```
Comment (repodiac) [inline train.py:498]: nope, apex is only loaded / required if the respective setting is active in the config as you can see

**Comment by repodiac** [comment]
> I just try to find a better solution. If people like to use APEX, we can serve as an option but if they don't we need to get rid of its 
> overhead. So I suggest having a train_apex.py and an option in setup.py to enable APEX if this is what the user desires.

I can see that APEX is another dependency and I am not sure if it only works with RTX GPU's (or the same architecture at least). So regarding installation behaviour I would also support you point with a optimized setup.py here - however, a pip installation would be best (as I also don't like the setup.py necessary as yet, in general). So we could create specific wheels, for example.

> In your current PR, even if you don't use APEX, still you need to install it since we import it anyway. Maybe one easier option is > to import APEX if it is available in the system.

No. I commented the respective line of code with a comment above. It is only loaded/active if the option is used in the config file.

**Comment by erogol** [comment]
I merged your PR to dev you can see your commit here. Thx and sorry for the delay. 

https://github.com/mozilla/TTS/pull/415

**Comment by repodiac** [comment]
Great. Thanks a lot :+1: Hopefully it deems useful to somebody :)

**Comment by erogol** [comment]
And just now PyTorch released native AMP support as I was afraid :) 

https://pytorch.org/blog/pytorch-1.6-released/

**Comment by repodiac** [comment]
I also read that recently. BUT: they have only begun to replace apex AMP. It is not fully replacable yet.

I checked their blog and docs regarding the new AMP feature in pytorch 1.6 - to me it seems they have done something similar as O1 optimization level (as in the pull request). But more optimized FP16 code (as in O2 or fully FP16 as in O3) is not available with pytorch amp as I see it.

https://nvidia.github.io/apex/amp.html

So do you plan stopping the merge into the master branch now? I would still think it is useful (as it is implemented and working now) at least until maybe the next pytorch release, where opt level O2+ is enabled. As I mentioned once, O2 did not work with your code - so pytorch amp won't work out of the box here, neither... I suppose.

**Comment by erogol** [comment]
If I have time I'll try Pytorch amp and see before mergin dev to master. But I'll edit on top of your PR.

**Comment by thorstenMueller** [comment]
I tried a training on latest dev branch but receive following error:
```
CUDA_VISIBLE_DEVICES="0" python mozilla_voice_tts/bin/train_tts.py --config_path /home/thorsten/___dev/tts/thorstenMueller/TTS_recipes/thorsten/model-config.json 
2020-08-07 15:35:55.737112: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/lib/x86_64-linux-gnu/:/usr/local/cuda/lib64:/usr/local/cuda-9.0/lib64:/usr/lib/x86_64-linux-gnu/:/usr/local/cuda/lib64:/usr/local/cuda-9.0/lib64:
2020-08-07 15:35:55.737144: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
 > Using CUDA:  True
 > Number of GPUs:  1
Traceback (most recent call last):
  File "mozilla_voice_tts/bin/train_tts.py", line 654, in <module>
    if c.apex_amp_level == 'O1':
AttributeError: 'AttrDict' object has no attribute 'apex_amp_level'
```

Did anyone see a similar error?

**Comment by thorstenMueller** [comment]
Sorry. I didn't really think about it. After crosschecking config file i found out that i just forgot property '"**apex_amp_level": null,**' in config.json.

**Pull request closed without merging**